packages/nimble-components/src/modal/index.ts

```diff
@@ -0,0 +1,143 @@
+     */
+    public get open(): boolean {
+        return this.state !== ModalState.closed;
+    }
+
+    private _state: ModalState = ModalState.closed;
+    private resolveShow?: (reason: CloseReason | UserDismissed) => void;
+
+    /**
+     * Opens the Modal
+     * @returns Promise that is resolved when the Modal is closed. The value of the resolved Promise is the reason value passed to the close() method, or UserDismissed if closed via the ESC key.
+     */
+    public async show(): Promise<CloseReason | UserDismissed> {
+        if (
+            this.state === ModalState.open
+            || this.state === ModalState.opening
+        ) {
+            throw new Error('Already open or opening');
+        }
+        this.state = ModalState.opening;
+        this.startOpening();
+        return new Promise((resolve, _reject) => {
+            this.resolveShow = resolve;
+        });
+    }
+
+    /**
+     * Closes the Modal
+     * @param reason An optional value indicating how/why the Modal was closed.
+     */
+    public close(reason: CloseReason): void {
+        if (this.state !== ModalState.open) {
+            throw new Error('Not open');
+        }
+        this.state = ModalState.closing;
+        this.startClosing(reason);
+    }
+
+    /**
+     * @internal
+     */
+    public cancelHandler(event: Event): boolean {
+        if (this.preventDismiss) {
+            event.preventDefault();
+        } else {
+            this.state = ModalState.closing;
+            this.startClosing(UserDismissed);
+        }
+        return true;
+    }
+
+    /**
+     * @internal
+     */
+    public closeHandler(): void {
+        if (this.resolveShow) {
+            // If
+            // - the browser implements dialogs with the CloseWatcher API, and
+            // - the user presses ESC without first interacting with the dialog (e.g. clicking, scrolling),
+            // the cancel event is not fired, but the close event still is, and the dialog just closes.
+            this.doResolveShow(UserDismissed);
+        }
+    }
+
+    // Derived classes can override this, but should not call it directly (except from the override).
+    protected startOpening(): void {
+        this.finishOpening();
+    }
+
+    // Overrides must call the base implementation, so that state gets updated.
+    protected finishOpening(): void {
+        this.state = ModalState.open;
+        this.dialogElement.showModal();
+    }
+
+    // Derived classes can override this, but should not call it directly (except from the override).
+    protected startClosing(reason: CloseReason | UserDismissed): void {
+        this.finishClosing(reason);
+    }
+
+    // Overrides must call the base implementation, so that state gets updated.
+    protected finishClosing(reason: CloseReason | UserDismissed): void {
+        this.state = ModalState.closed;
+        this.dialogElement.close();
+        this.doResolveShow(reason);
+    }
```

Comment:
I think the issues here for needing a buunch of overrides are that the Modal class is trying to implement a specific workflow, where the transition is preset so i.e. startClosing immediately calls finishClosing. Instead there should be lifecycle hooks that each subclass is implementing.

startOpening is abstract and in the dialog implementation iit immediately calls finishOpeniing while in the drawer implementation it does its animations beforee calling finishClosing.

Reply:
I made `startOpening/Closing` abstract. It doesn't really change much, other than moving the default implementations to `Dialog`. I guess it technically reduces the number of overrides by 2 (if you don't count the abstract method implementations). That's all you wanted me to change, right?